### ISSUE TITLE
[Question] Is the axis order optimization correct?

### ISSUE THREAD

**Issue by embeddedt** [Steveplays28/noisium#23]
PalettedContainers store their data in YZX order. This can be confirmed by looking at `PalettedContainer.Strategy#getIndex` (mojmap). However, the axis order optimization in Noisium appears to be [iterating in XZY order](https://github.com/Steveplays28/noisium/blob/76579299ef01ff0c169c3cacef864e7b0057dc37/common/src/main/java/io/github/steveplays28/noisium/mixin/ChunkSectionMixin.java#L31-L33). I don't know if this actually reduces performance with the size of a modern CPU cache, but it seems unintended regardless. Ideally, I think you would want the loop to also be in YZX order (i.e. increment X, then Y, then Z), so that you are setting blocks at sequentially increasing indices. The current loop increments Y first before the other coordinates and won't do that.

**Comment by Steveplays28** [comment]
Hi embeddedt, thanks for bringing this up. Seems like neither my method nor vanilla's method are best here lmao.
Vanilla simply does XYZ and Noisium does XZY, while the palette indeed does YZX.

The original thought process was that there'd be more data to swap on the X axis and the Z axis, so that'd be done first, to improve performance with slower memory. That's probably questionable though, what you're saying sounds like a better idea.
Let me know what you think about this

It's easy enough to change so I'll do some profiling on all three options

EDIT: I think I misread the part about the iteration order a bit, you're saying Y gets incremented first which isn't what I wanted to achieve when I wrote this, I don't think. :rofl:
YZX is probably what I wanted to do, would check out with my explanation above, that way Z and X are incremented first, and Y last

**Comment by Steveplays28** [comment]
So, just to double check, the fix would be to swap Y with X:
```java
for (int posY = 0; posY < sliceSize; ++posY) {
	for (int posZ = 0; posZ < sliceSize; ++posZ) {
		for (int posX = 0; posX < sliceSize; ++posX) {
			palettedContainer.swapUnsafe(posX, posY, posZ, biomeSupplier.getBiome(x + posX, y + posY, z + posZ, sampler));
		}
	}
}
```
that way X and Z positions are incremented before Y positions.

Haven't benchmarked yet though

**Comment by embeddedt** [comment]
Yes, I don't know if the performance difference will be measurable, but it makes more sense to do it this way anyways, given the structure of the data in memory.

**Comment by Steveplays28** [comment]
Ok, awesome. I'm also curious if it'll be measurable, doing some (short) benchmarks now.
Thanks again for spotting this.

Will let you know of the results of the benchmarks when I'm done

**Comment by Steveplays28** [comment]
Maybe it's random variation, but I'm seeing a minor difference on my end (<0.5% improvement).
Will push this change and include it in the next patch release!

**Comment by Steveplays28** [comment]
Just pushed the fix for this in [`v2.0.2`](https://github.com/Steveplays28/noisium/releases/latest).
